### PR TITLE
ethdb/remotedb, cmd: add support for remote (readonly) databases

### PR DESCRIPTION
**Pull request by holiman**
I'm not sure if this PR belongs in geth -- it's one of those PRs which doesn't _have_ to be in mainline geth, since it's not useful for the bulk of users. 
This is useful for investigating remote nodes. In many cases when something odd happens, we have to ask people to do various tasks, looking up db values or checking things over rpc. 

Alternatively, we make a special branch with some investigative tools and ask them to compile + redeploy. 

Since a while back, we added `debug_dbGet` to the rpc namespace, which allows us to just wrap the remote rpc endpoint as a database (it'll be readonly, since we decided not to include deletion/modification). 

Examples

Attaching to a local rpc node and getting the metadata: 
```
$ ./build/bin/geth --remotedb "ipc:///home/user/.ethereum/sepolia/geth.ipc" db metadata
INFO [05-09|13:37:50.715] Maximum peer count                       ETH=50 LES=0 total=50
INFO [05-09|13:37:50.715] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [05-09|13:37:50.716] Set global gas cap                       cap=50,000,000
INFO [05-09|13:37:50.716] Allocated cache and file handles         database=/home/user/.ethereum/geth/chaindata cache=512.00MiB handles=262,144 readonly=true
INFO [05-09|13:37:50.724] Opened ancient database                  database=/home/user/.ethereum/geth/chaindata/ancient readonly=true
+-------------------------+--------------------------------------------------------------------+
|          FIELD          |                               VALUE                                |
+-------------------------+--------------------------------------------------------------------+
| databaseVersion         | 8 (0x8)                                                            |
| headBlockHash           | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| headFastBlockHash       | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| headHeaderHash          | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| headBlock.Hash          | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| headBlock.Root          | 0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544 |
| headBlock.Number        | 0 (0x0)                                                            |
| headHeader.Hash         | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| headHeader.Root         | 0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544 |
| headHeader.Number       | 0 (0x0)                                                            |
| frozen                  | 0 items                                                            |
| lastPivotNumber         | <nil>                                                              |
| len(snapshotSyncStatus) | 0 bytes                                                            |
| snapshotGenerator       | Done: true, Accounts: 0,                                           |
|                         | Slots: 0, Storage: 0, Marker:                                      |
| snapshotDisabled        | false                                                              |
| snapshotJournal         | 34 bytes                                                           |
| snapshotRecoveryNumber  | <nil>                                                              |
| snapshotRoot            | 0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544 |
| txIndexTail             | <nil>                                                              |
| fastTxLookupLimit       | <nil>                                                              |
+-------------------------+--------------------------------------------------------------------+

```
Attaching to a local ipc endpoint and fetching the snapshot root
```
$ ./build/bin/geth --remotedb "ipc:///home/user/.ethereum/sepolia/geth.ipc" db get SnapshotRoot
INFO [05-09|13:50:50.530] Maximum peer count                       ETH=50 LES=0 total=50
INFO [05-09|13:50:50.530] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [05-09|13:50:50.531] Set global gas cap                       cap=50,000,000
INFO [05-09|13:50:50.531] Allocated cache and file handles         database=/home/user/.ethereum/geth/chaindata cache=512.00MiB handles=262,144 readonly=true
INFO [05-09|13:50:50.535] Opened ancient database                  database=/home/user/.ethereum/geth/chaindata/ancient readonly=true
key 0x536e617073686f74526f6f74: 0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544
```
Dumping the snapshot state
```
[user@work go-ethereum]$ ./build/bin/geth --remotedb "ipc:///home/user/.ethereum/sepolia/geth.ipc" snapshot dump
INFO [05-09|13:51:49.470] Maximum peer count                       ETH=50 LES=0 total=50
INFO [05-09|13:51:49.470] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [05-09|13:51:49.472] Set global gas cap                       cap=50,000,000
INFO [05-09|13:51:49.472] Allocated cache and file handles         database=/home/user/.ethereum/geth/chaindata cache=512.00MiB handles=262,144 readonly=true
INFO [05-09|13:51:49.477] Opened ancient database                  database=/home/user/.ethereum/geth/chaindata/ancient readonly=true
INFO [05-09|13:51:49.477] State dump configured                    block=0 hash=0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 skipcode=false skipstorage=false start=0x0000000000000000000000000000000000000000000000000000000000000000 limit=0
INFO [05-09|13:51:49.478] Snapshot dumping started                 root=d7f897..0f0544
{"root":"0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544"}
{"balance":"466215000000000000000","nonce":0,"root":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","codeHash":"0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470","key":"0x000388c5ba62b0e7342687d94b0e03b772aa4ab7c08f13fe3fa9f9d0a3153e05"}
{"balance":"100000000000000000000000","nonce":0,"root":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","codeHash":"0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470","key":"0x0004204188718653cd7e50f3fd51a820db66112517ca190c637e7cdd80782d56"}
{"balance":"2000000000000000000000","nonce":0,"root":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","codeHash":"0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470","key":"0x0004a390953f116afb00f89fbedb2f8e77297e4e7e1749e2ef0e32e17808e4ad"}
{"balance":"910000000000000000000","nonce":0,"root":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","codeHash":"0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470","key":"0x0005bd66f60e5b954e1af93ded1b02cb575ff0ed6d9241797eff7576b0bf0637"}
{"balance":"206764195000000000000000","nonce":0,"root":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","codeHash":"0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470","key":"0x00074ae55b1d051a42ba19c0cc51f93eae4ab2db5a2eebc31a7b89cb0bb88039"}
{"balance":"16011846000000000000000","nonce":0,"root":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","codeHash":"0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470","key":"0x0016c9db9bb545a03425e300f3ee72bae098110336dd3eaf48c20a2e5b6865fc"}
```
Verifying the snapshot state
```
$ ./build/bin/geth --remotedb "ipc:///home/user/.ethereum/sepolia/geth.ipc" snapshot verify-state
INFO [05-09|13:52:24.602] Maximum peer count                       ETH=50 LES=0 total=50
INFO [05-09|13:52:24.602] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [05-09|13:52:24.603] Set global gas cap                       cap=50,000,000
INFO [05-09|13:52:24.604] Allocated cache and file handles         database=/home/user/.ethereum/geth/chaindata cache=512.00MiB handles=262,144 readonly=true
INFO [05-09|13:52:24.612] Opened ancient database                  database=/home/user/.ethereum/geth/chaindata/ancient readonly=true
INFO [05-09|13:52:24.614] Iterating state snapshot                 accounts=0 slots=0 elapsed="304.4µs"
INFO [05-09|13:52:24.869] Iterated snapshot                        accounts=8893 elapsed=255.693ms
INFO [05-09|13:52:24.869] Verified the state                       root=d7f897..0f0544
INFO [05-09|13:52:24.869] Checking dangling snapshot disk storage 
INFO [05-09|13:52:24.869] Verified the snapshot disk storage       time="26.632µs" err=nil
INFO [05-09|13:52:24.869] Checking dangling snapshot difflayer journalled storage 
INFO [05-09|13:52:24.869] Verified the snapshot journalled storage time="16.572µs"
```

### CODE REVIEW

File: ethdb/remotedb/remotedb.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+func (db *Database) Compact(start []byte, limit []byte) error {
+	return nil
+}
+
+func (db *Database) NewSnapshot() (ethdb.Snapshot, error) {
+	panic("not supported")
+}
+
+func (db *Database) Close() error {
+	return nil
```
Comment (s1na) [inline ethdb/remotedb/remotedb.go:138]:
```suggestion
        db.remote.Close()
	return nil
```

**Review by s1na — APPROVED** [review]
This is very cool!

**Comment by holiman** [comment]
Triage decision: 
- Add support for ancient data too.

**Comment by holiman** [comment]
Added support for looking up ancient data now. With that in place, the `metadata` spits out more information:
```
$ geth --remotedb "ipc:///home/user/.ethereum/sepolia/geth.ipc" db metadata
INFO [05-10|13:47:54.786] Maximum peer count                       ETH=50 LES=0 total=50
INFO [05-10|13:47:54.786] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [05-10|13:47:54.787] Set global gas cap                       cap=50,000,000
INFO [05-10|13:47:54.787] Using remote db                          url=ipc:///home/user/.ethereum/sepolia/geth.ipc
+-------------------------+--------------------------------------------------------------------+
|          FIELD          |                               VALUE                                |
+-------------------------+--------------------------------------------------------------------+
| databaseVersion         | 8 (0x8)                                                            |
| headBlockHash           | 0x35afa8a458443f5e85905eaa796713aca4237381e97ec34c89cd9d8d717b7a66 |
| headFastBlockHash       | 0x35afa8a458443f5e85905eaa796713aca4237381e97ec34c89cd9d8d717b7a66 |
| headHeaderHash          | 0x35afa8a458443f5e85905eaa796713aca4237381e97ec34c89cd9d8d717b7a66 |
| headBlock.Hash          | 0x35afa8a458443f5e85905eaa796713aca4237381e97ec34c89cd9d8d717b7a66 |
| headBlock.Root          | 0x059c6336aa43cd4471385a9b4f37004139333dd790db39e54f387901aa4d27a9 |
| headBlock.Number        | 149652 (0x24894)                                                   |
| headHeader.Hash         | 0x35afa8a458443f5e85905eaa796713aca4237381e97ec34c89cd9d8d717b7a66 |
| headHeader.Root         | 0x059c6336aa43cd4471385a9b4f37004139333dd790db39e54f387901aa4d27a9 |
| headHeader.Number       | 149652 (0x24894)                                                   |
| frozen                  | 59653 items                                                        |
| lastPivotNumber         | 96168 (0x177a8)                                                    |
| len(snapshotSyncStatus) | 299 bytes                                                          |
| snapshotGenerator       | Done: true, Accounts: 0,                                           |
|                         | Slots: 0, Storage: 0, Marker:                                      |
| snapshotDisabled        | true                                                               |
| snapshotJournal         | 11370 bytes                                                        |
| snapshotRecoveryNumber  | <nil>                                                              |
| snapshotRoot            | 0xcf36401ce2143625f9823620740e59efa93445451acc7cc1e8b621bbc6dab95b |
| txIndexTail             | 0 (0x0)                                                            |
| fastTxLookupLimit       | 2350000 (0x23dbb0)                                                 |
+-------------------------+--------------------------------------------------------------------+
```

**Review by karalabe — APPROVED** [review]
SGTM

**Comment by karalabe** [comment]
Myeah, I think that's a bug :) Ping @holiman.

Thanks @hsyodyssey

**Pull request merged**